src-tauri/src/constructors/init/init_edge_map.rs

```diff
@@ -1,37 +1,72 @@
-use crate::analytics::aux_data_structures::neighbor_info::{Neighbor, NeighborInfo};
+use app::protobufs::{Neighbor, NeighborInfo};
 use std::collections::HashMap;
 
-// Create a hashmap of edges and their SNR values from a vector of NeighborInfo packets.
-// The hashmap is of form (node_id_1, node_id_2) -> (SNR, timestamp).
-pub fn init_edge_map(packets: Vec<NeighborInfo>) -> HashMap<(u32, u32), (f64, u64)> {
+/*
+* Create a hashmap of edges and their SNR values from a vector of NeighborInfo packets.
+* When we have a conflict between two nodes reporting the same edge, we take the most recent
+* timestamp as fact (if the most recently updated node does not report the edge, we remove it).
+* @returns a hashmap is of form (node_id_1, node_id_2) -> (SNR, timestamp).
+ */
+pub fn init_edge_map(neighbors: HashMap<u32, NeighborInfo>) -> HashMap<(u32, u32), (f64, u64)> {
     let mut snr_hashmap = HashMap::<(u32, u32), (f64, u64)>::new();
-    for node_packet in packets {
-        let node_id = node_packet.id;
-        for neighbor_packet in node_packet.neighbors {
-            let neighbor_id = neighbor_packet.id;
-            // Check if the edge is already in the hashmap
-            let edge_exists = snr_hashmap.contains_key(&as_key(node_id, neighbor_id));
-            if edge_exists {
-                let edge_snr_time = snr_hashmap.get(&as_key(node_id, neighbor_id)).unwrap();
-                let hashmap_timestamp = edge_snr_time.1;
-                // Replace the edge if the new timestamp is more recent (larger)
-                if hashmap_timestamp < neighbor_packet.timestamp {
+    for (_, neighbor_packet) in &neighbors {
+        let node_1 = neighbor_packet.node_id;
+        for neighbor in &neighbor_packet.neighbors {
+            let node_2 = neighbor.node_id;
+            snr_hashmap.insert(
+                as_key(node_1, node_2),
+                (neighbor.snr as f64, neighbor.rx_time as u64),
+            );
+            // If the opposite neighbor is found on a recent packet, we take the most recent timestamp as fact.
+            let opposite_neighbor = get_other_edge_side_packet(node_2, node_1, &neighbors);
+            if opposite_neighbor.is_some() {
+                let opposite_neighbor = opposite_neighbor.unwrap();
+                if neighbor.rx_time > opposite_neighbor.rx_time {
                     snr_hashmap.insert(
-                        as_key(node_id, neighbor_id),
-                        (neighbor_packet.snr, neighbor_packet.timestamp),
+                        as_key(node_1, node_2),
+                        (neighbor.snr as f64, neighbor.rx_time as u64),
+                    );
+                } else {
+                    snr_hashmap.insert(
+                        as_key(node_1, node_2),
+                        (opposite_neighbor.snr as f64, opposite_neighbor.rx_time as u64),
```

Comment:
Nit: Could this if/else be done with a ternary or conditionally-assigned variable? Not sure it's worth having a completely different if/else block if only one arg is changing

Reply:
Replaced with ternary (as close as Rust gets)